test/test_model.py

```diff
@@ -212,6 +212,178 @@ def check_remote_units():
             ['relation-list', '-r', '4', '--format=json'],
         ])
 
+    def test_base_status_instance_raises(self):
+        with self.assertRaises(TypeError):
+            op.model.Status('test')
+
+    def test_local_set_valid_unit_status(self):
+        self.backend = op.model.ModelBackend()
+        self.model = op.model.Model('myapp/0', ['db0', 'db1', 'db2'], self.backend)
+
+        test_cases = (
+            (
+                op.model.ActiveStatus('Green'),
+                lambda: fake_script(self, 'status-set', 'exit 0'),
+                lambda: self.assertEqual(fake_script_calls(self, True), [['status-set', '--application=False', 'active', 'Green']]),
+            ),
+            (
+                op.model.MaintenanceStatus('Yellow'),
+                lambda: fake_script(self, 'status-set', 'exit 0'),
+                lambda: self.assertEqual(fake_script_calls(self, True), [['status-set', '--application=False', 'maintenance', 'Yellow']]),
+            ),
+            (
+                op.model.BlockedStatus('Red'),
+                lambda: fake_script(self, 'status-set', 'exit 0'),
+                lambda: self.assertEqual(fake_script_calls(self, True), [['status-set', '--application=False', 'blocked', 'Red']]),
+            ),
+            (
+                op.model.WaitingStatus('White'),
+                lambda: fake_script(self, 'status-set', 'exit 0'),
+                lambda: self.assertEqual(fake_script_calls(self, True), [['status-set', '--application=False', 'waiting', 'White']]),
+            ),
+        )
+
+        for target_status, setup_tools, check_tool_calls in test_cases:
+            setup_tools()
+
+            self.model.unit.status = target_status
+
+            self.assertEqual(self.model.unit.status, target_status)
+
+            check_tool_calls()
+
+    def test_local_set_valid_app_status(self):
+        self.backend = op.model.ModelBackend()
+        self.model = op.model.Model('myapp/0', ['db0', 'db1', 'db2'], self.backend)
+
+        fake_script(self, 'is-leader', 'echo true')
+
+        # TODO: update after is-leader caching optimization is merged
+        # this will break those test cases as is-leader hook tool invocation
+        # will depend on time.
+
+        test_cases = (
+            (
+                op.model.ActiveStatus('Green'),
+                lambda: fake_script(self, 'status-set', 'exit 0'),
+                lambda: self.assertEqual(fake_script_calls(self, True), [
+                    ['is-leader', '--format=json'],
+                    ['status-set', '--application=True', 'active', 'Green'],
+                    ['is-leader', '--format=json'],
+                ]),
```

Comment:
Since I don't think we actually care to test the `is-leader` calls, how about instead of doing `assertEqual` here, use `assertIn` to only confirm that the `['status-set', '--application=True', 'active', 'Green']` record is in the call list.  That would future-proof these tests against the `is-leader` timed caching, as well as making the test a bit less noisy.

Reply:
Agreed, anything extra would raise an exception as we do not create any other fake hook tools.

Thanks for the suggestion.